### PR TITLE
Support W3C tracing + example.

### PR DESCRIPTION
**Pull request by artursouza**
# Description

Adds support for W3C tracing via OpenTelemetrySDK.
Adds example for W3C tracing.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #298 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation

### CODE REVIEW

File: examples/src/main/java/io/dapr/examples/tracing/DemoServiceController.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Map;
+import java.util.TimeZone;
+
+/**
+ * SpringBoot Controller to handle service invocation.
+ *
+ * <p>Instrumentation is handled in {@link io.dapr.springboot.OpenTelemetryInterceptor}.
+ */
+@RestController
+public class DemoServiceController {
```
Comment (Haishi2016) [inline examples/src/main/java/io/dapr/examples/tracing/DemoServiceController.java:30]: DemoServiceController is defined in both invoke sample and tracing sample. This leads to exception when the sample is launched:

←[94;1m== APP == 2020-08-17 12:56:44,930 {HH:mm:ss.SSS} [main] INFO  io.dapr.examples.tracing.DemoService - Starting DemoService on DESKTOP-HRKK1LN with PID 19900 (C:\Users\hbai\go\src\github.com\dapr\java-sdk\examples\target\dapr-java-sdk-examples-exec.jar started by hbai in C:\Users\hbai\go\src\github.com\dapr\java-sdk\examples)
←[0m
←[94;1m== APP == 2020-08-17 12:56:44,932 {HH:mm:ss.SSS} [main] INFO  io.dapr.examples.tracing.DemoService - No active profile set, falling back to default profiles: default
←[0m
←[94;1m== APP == 2020-08-17 12:56:45,386 {HH:mm:ss.SSS} [main] WARN  o.s.b.w.s.c.AnnotationConfigServletWebServerApplicationContext - Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [io.dapr.springboot.DaprApplication]; nested exception is org.springframework.context.annotation.ConflictingBeanDefinitionException: Annotation-specified bean name 'demoServiceController' for bean class [io.dapr.examples.tracing.DemoServiceController] conflicts with existing, non-compatible bean definition of same name and class [io.dapr.examples.invoke.http.DemoServiceController]
←[0m
←[94;1m== APP == 2020-08-17 12:56:45,724 {HH:mm:ss.SSS} [background-preinit] WARN  o.s.h.c.j.Jackson2ObjectMapperBuilder - For Jackson Kotlin classes support please add "com.fasterxml.jackson.module:jackson-module-kotlin" to the classpath
←[0m
←[94;1m== APP == 2020-08-17 12:56:45,753 {HH:mm:ss.SSS} [main] ERROR o.s.boot.SpringApplication - Application run failed
←[0m
←[94;1m== APP == org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [io.dapr.springboot.DaprApplication]; nested exception is org.springframework.context.annotation.ConflictingBeanDefinitionException: Annotation-specified bean name 'demoServiceController' for bean class [io.dapr.examples.tracing.DemoServiceController] conflicts with existing, non-compatible bean definition of same name and class [io.dapr.examples.invoke.http.DemoServiceController]
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:184)
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:325)
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:242)
←[0m
←[94;1m== APP ==        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:275)
←[0m
←[94;1m== APP ==        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:95)
←[0m
←[94;1m== APP ==        at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:706)
←[0m
←[94;1m== APP ==        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:532)
←[0m
←[94;1m== APP ==        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141)
←[0m
←[94;1m== APP ==        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747)
←[0m
←[94;1m== APP ==        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
←[0m
←[94;1m== APP ==        at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
←[0m
←[94;1m== APP ==        at io.dapr.springboot.DaprApplication.start(DaprApplication.java:24)
←[0m
←[94;1m== APP ==        at io.dapr.examples.tracing.DemoService.main(DemoService.java:43)
←[0m
←[94;1m== APP ==        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
←[0m
←[94;1m== APP ==        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
←[0m
←[94;1m== APP ==        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
←[0m
←[94;1m== APP ==        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
←[0m
←[94;1m== APP ==        at io.dapr.springboot.DaprMainApplication.main(DaprMainApplication.java:29)
←[0m
←[94;1m== APP ==        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
←[0m
←[94;1m== APP ==        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
←[0m
←[94;1m== APP ==        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
←[0m
←[94;1m== APP ==        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
←[0m
←[94;1m== APP ==        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
←[0m
←[94;1m== APP ==        at org.springframework.boot.loader.Launcher.launch(Launcher.java:109)
←[0m
←[94;1m== APP ==        at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
←[0m
←[94;1m== APP ==        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88)
←[0m
←[94;1m== APP == Caused by: org.springframework.context.annotation.ConflictingBeanDefinitionException: Annotation-specified bean name 'demoServiceController' for bean class [io.dapr.examples.tracing.DemoServiceController] conflicts with existing, non-compatible bean definition of same name and class [io.dapr.examples.invoke.http.DemoServiceController]
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ClassPathBeanDefinitionScanner.checkCandidate(ClassPathBeanDefinitionScanner.java:349)
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ClassPathBeanDefinitionScanner.doScan(ClassPathBeanDefinitionScanner.java:287)
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ComponentScanAnnotationParser.parse(ComponentScanAnnotationParser.java:132)
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ConfigurationClassParser.doProcessConfigurationClass(ConfigurationClassParser.java:290)
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ConfigurationClassParser.processConfigurationClass(ConfigurationClassParser.java:245)
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:202)
←[0m
←[94;1m== APP ==        at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:170)
←[0m
←[94;1m== APP ==        ... 25 common frames omitted

File: examples/src/main/java/io/dapr/springboot/OpenTelemetryConfig.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import io.opentelemetry.trace.Tracer;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
+import org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@EnableAutoConfiguration(exclude = {DataSourceAutoConfiguration.class, HibernateJpaAutoConfiguration.class})
+public class OpenTelemetryConfig {
```
Comment (Haishi2016) [inline examples/src/main/java/io/dapr/springboot/OpenTelemetryConfig.java:23]: Do we want to integrate with Dapr configuration?

File: README.md [hunk 3]
```diff
@@ -140,6 +140,7 @@ Try the following examples to learn more about Dapr's Java SDK:
 * [PubSub with subscriber over Http](./examples/src/main/java/io/dapr/examples/pubsub/http)
 * [Binding with input over Http](./examples/src/main/java/io/dapr/examples/bindings/http)
 * [Actors over Http](./examples/src/main/java/io/dapr/examples/actors/http)
+* [Distributed tracing with OpenTelemetrySDK](./examples/src/main/java/io/dapr/examples/tracing)
```
Comment (shalabhms) [inline README.md:143]: nit: space, OpenTelemetry SDK

File: examples/src/main/java/io/dapr/examples/tracing/DemoServiceController.java (hunk 1 above)
Comment (artursouza) [reply examples/src/main/java/io/dapr/examples/tracing/DemoServiceController.java:30]: Nice catch. Let me fix this now.

File: examples/src/main/java/io/dapr/springboot/OpenTelemetryConfig.java (hunk 2 above)
Comment (artursouza) [reply examples/src/main/java/io/dapr/springboot/OpenTelemetryConfig.java:23]: I am not sure if I understand the proposal here. This is an example of how users can integrate with OpenTelemetry, but Dapr should not dictate that - this is 100% up to the application to decide how to initialize OpenTelemetry in their app.

File: README.md (hunk 3 above)
Comment (artursouza) [reply README.md:143]: Done.

**Comment by artursouza** [comment]
@shalabhms W3c in the SDK + doc + examples must be together. This is a breaking change and it affects other parts of the code. The only unrelated change is an organization of imports across the project. Tests also needed to be changed due to breaking changes in the SDK code. Please, review in chunks if needed.

**Comment by artursouza** [comment]
Updated with the fixes.

File: sdk/src/main/java/io/dapr/client/domain/InvokeServiceRequest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  void setBody(Object body) {
+    this.body = body;
+  }
+
+  public Map<String, String> getMetadata() {
+    return metadata;
+  }
+
+  void setMetadata(Map<String, String> metadata) {
+    this.metadata = metadata;
```
Comment (mukundansundar) [inline sdk/src/main/java/io/dapr/client/domain/InvokeServiceRequest.java:58]: unmodifiableMap ?

File: sdk/src/main/java/io/dapr/client/domain/InvokeBindingRequest.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  void setData(Object data) {
+    this.data = data;
+  }
+
+  public Map<String, String> getMetadata() {
+    return metadata;
+  }
+
+  void setMetadata(Map<String, String> metadata) {
+    this.metadata = metadata;
```
Comment (mukundansundar) [inline sdk/src/main/java/io/dapr/client/domain/InvokeBindingRequest.java:56]: unmodifiableMap?

File: sdk/src/main/java/io/dapr/client/domain/GetSecretRequest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  void setKey(String key) {
+    this.key = key;
+  }
+
+  public Map<String, String> getMetadata() {
+    return metadata;
+  }
+
+  void setMetadata(Map<String, String> metadata) {
+    this.metadata = metadata;
```
Comment (mukundansundar) [inline sdk/src/main/java/io/dapr/client/domain/GetSecretRequest.java:47]: unmodifiableMap? same for other requests?

File: sdk/src/main/java/io/dapr/client/domain/SaveStateRequest.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+package io.dapr.client.domain;
+
+import io.grpc.Context;
+
+import java.util.List;
+
+/**
+ * A request to save states to state store.
+ */
+public class SaveStateRequest {
```
Comment (mukundansundar) [inline sdk/src/main/java/io/dapr/client/domain/SaveStateRequest.java:15]: Does this not contain a metadata?

File: examples/src/main/java/io/dapr/examples/tracing/README.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  }
+///...
+}
+```
+
+The class knows the app id for the remote application. It uses `invokeService` method to invoke API calls on the service endpoint. The request object includes an instance of `io.grpc.Context` for the proper tracing headers to be propagated.
+ 
+Execute the follow script in order to run the InvokeClient example, passing two messages for the remote method:
+```sh
+dapr run --components-path ./components -- java -jar target/dapr-java-sdk-examples-exec.jar io.dapr.examples.tracing.InvokeClient 'message one' 'message two'
```
Comment (Haishi2016) [inline examples/src/main/java/io/dapr/examples/tracing/README.md:182]: for Windows, message one and message two should be wrapped by double quotes instead.

Comment (artursouza) [reply examples/src/main/java/io/dapr/examples/tracing/README.md:182]: done.

File: sdk/src/main/java/io/dapr/client/domain/SaveStateRequest.java (hunk 7 above)
Comment (artursouza) [reply sdk/src/main/java/io/dapr/client/domain/SaveStateRequest.java:15]: No. The class from proto also doesn't.

File: sdk/src/main/java/io/dapr/client/domain/GetSecretRequest.java (hunk 6 above)
Comment (artursouza) [reply sdk/src/main/java/io/dapr/client/domain/GetSecretRequest.java:47]: This method is not public on purpose. The builder classes take care of the unmodifiableMap.

File: sdk/src/main/java/io/dapr/client/domain/InvokeBindingRequest.java (hunk 5 above)
Comment (artursouza) [reply sdk/src/main/java/io/dapr/client/domain/InvokeBindingRequest.java:56]: this.metadata = metadata == null ? null : Collections.unmodifiableMap(metadata);

File: sdk/src/main/java/io/dapr/client/domain/InvokeServiceRequest.java (hunk 4 above)
Comment (artursouza) [reply sdk/src/main/java/io/dapr/client/domain/InvokeServiceRequest.java:58]: This method is not public on purpose. The builder classes take care of the unmodifiableMap.

**Comment by mukundansundar** [comment]
Let us discuss this tomorrow before merging or not.

**Comment by artursouza** [comment]
Updated the PR with Multi PubSub.

**Pull request merged**